Feat/replace academic calendar with restapi

timGalk opened the pull request:
This pull request introduces significant changes to the academic calendar feature, including a migration from GraphQL to REST APIs, the addition of new models for academic calendar and day swaps, and various updates to accommodate the new data structures. It also includes minor dependency updates and cleanup in the `pubspec.yaml` file.

### Migration to REST API and Data Model Updates:
* Replaced GraphQL-based `academic_calendar_repo.dart` with a REST-based implementation, introducing `AcademicCalendarWithSwaps` to combine calendar data and day swaps. Added REST endpoints for fetching academic calendar and day swap data.
* Removed the GraphQL query file `getAcademicCalendar.graphql` as it is no longer needed.

### New Models for Academic Calendar and Day Swaps:
* Added `AcademicCalendar` and `AcademicCalendarResponse` models to represent the academic calendar data.
* Introduced `DaySwap` and `DaySwapList` models, along with `DaySwapData` and `DaySwapResponse` for handling day swap information. [[1]](diffhunk://#diff-879b52b73b21ac314c9b4321344555132bccff6b70849cf04fcb53425d366706R1-R25) [[2]](diffhunk://#diff-46fabfbb0f8945530739dc6b636feb5ef4edd95eb9bb43f1aa71433a25429306R1-R26)

### Refactoring for New Data Structures:
* Updated various extensions, widgets, and utilities to use `AcademicCalendarWithSwaps` instead of the old GraphQL-based types. This includes `academic_calendar_extensions.dart`, `academic_week_exception.dart`, and widgets like `ExamSessionCountdown` and `Greeting`. [[1]](diffhunk://#diff-9de51e26f447c87597ca9c5a345221725b50876a19aa6974295ec5064087ade5R3-R14) [[2]](diffhunk://#diff-9de51e26f447c87597ca9c5a345221725b50876a19aa6974295ec5064087ade5L39-R45) [[3]](diffhunk://#diff-76a406ba698b4ea5706a23c5e62a235d502c7daf927a724c7ff50290508a48c1L10-R19) [[4]](diffhunk://#diff-fa0977bd66502e41be62dbd769f8ff166421d89a151a6c83ce15cc49350902f0L15-R15) [[5]](diffhunk://#diff-9d74e40eadd575292502f5687e3c3fb20052e5882e704c161dd7407942b6c15aL21-R26)

### Dependency and Build Configuration Changes:
* Updated the `com.android.application` plugin version in `android/settings.gradle` from `8.3.2` to `8.6.0`.
* Removed the `svg_optimizer` dependency and its associated configuration in `pubspec.yaml`. [[1]](diffhunk://#diff-8b7e9df87668ffa6a04b32e1769a33434999e54ae081c52e5d943c541d4c0d25L155) [[2]](diffhunk://#diff-8b7e9df87668ffa6a04b32e1769a33434999e54ae081c52e5d943c541d4c0d25L174-L175)

mikolaj-jalocha commented on android/settings.gradle:
Was this necessary?

mikolaj-jalocha commented on lib/features/academic_calendar/model/academic_calendar_extensions.dart:
I think we can exclude "Rest" appendix, since finally every data that've got will be fetched the same way

mikolaj-jalocha commented on lib/features/academic_calendar/model/academic_calendar_rest_model.dart:
```diff
@@ -0,0 +1,29 @@
+  const factory AcademicCalendarRestData({
+    required int id,
+    required String name,
+    @JsonKey(name: "semesterStartDate") required DateTime semesterStartDate,
+    @JsonKey(name: "examSessionStartDate") required DateTime examSessionStartDate,
+    @JsonKey(name: "examSessionLastDate") required DateTime examSessionLastDate,
```
Are JsonKey name annotations necessary here? Name of the variables are the same as in annotation.

mikolaj-jalocha commented on lib/features/academic_calendar/model/day_swap_rest_model.dart:
```diff
@@ -0,0 +1,26 @@
+    required int id,
+    required int academicCalendarId,
+    required DateTime date,
+    @JsonKey(name: "changedWeekday") required String changedWeekday,
+    @JsonKey(name: "changedDayIsEven") required bool changedDayIsEven,
```
Same here

mikolaj-jalocha commented on lib/features/academic_calendar/repository/academic_calendar_repo.dart:
```diff
@@ -1,23 +1,53 @@
+  final calendarData = responses[0] as AcademicCalendarRestResponse;
+  final daySwaps = responses[1] as DaySwapRestResponse;
+
+  return AcademicCalendarData(calendarData: calendarData.data.first, daySwaps: daySwaps.data.toIList());
```
Unlikely to happen imo, but maybe worth implementing for the quality of sleep :)

mikolaj-jalocha commented on pubspec.yaml:
```diff
@@ -131,6 +131,7 @@ dependencies:
   flutter_hooks: ^0.21.2
   audioplayers: ^6.2.0
 
+  http: any
 dev_dependencies:
```
What's that? 🧐

mikolaj-jalocha reviewed (changes requested):
Overall everything looks good. Logic same as before, data correctly fetched and passed down. Below're suggestions and questions of mine:

1. Why changing `android.application `version and what'is `http `in `yaml`?
2. I'd totally omit `Rest `appending especially in naming of classes. It's implementation detail, others shouldn't know about it :) Besides, all data we fetch comes from REST api. 
3. This probably could wait till all rest endpoints will be implemented, but I personally adopted a strategy where object that serves as DTO are marked with `Data `appendix, anm objects that serves as model classes in UI are without it. Worth considering here.

simon-the-shark commented on lib/features/academic_calendar/repository/academic_calendar_repo.dart:
```diff
@@ -1,23 +1,53 @@
+    ),
+  ]);
+  final calendarData = responses[0] as AcademicCalendarRestResponse;
+  final daySwaps = responses[1] as DaySwapRestResponse;
```
can we instead of casting with `as` just do 

```dart
final calendarData = AcademicCalendarRestResponse.fromJson(responses[0] as Map<String, dynamic>);
  final daySwaps = DaySwapRestResponse.fromJson(responses[1] as Map<String, dynamic>);
```

simon-the-shark reviewed (commented):
yeah pretty much Mikołaj pointed out all needed adjustments. I have only one extra note. But good job here! 🚀 It looks really good overall.

timGalk commented:
This pull request introduces significant changes to the academic calendar feature, transitioning from a GraphQL-based implementation to a REST-based one, and restructuring the data model to include day swaps alongside academic calendar data. Additionally, it updates dependencies and removes unused tools. Below are the most important changes grouped by theme:

### Transition to REST API and Data Model Restructuring:
* **New Models for REST API Integration**: Added `AcademicCalendarResponse`, `AcademicCalendar`, `DaySwapResponse`, and `DaySwapData` classes to handle REST API responses using the `freezed` package. These replace the previous GraphQL models. (`lib/features/academic_calendar/model/academic_calendar.dart` [[1]](diffhunk://#diff-6e5add880fa8e6be726f9a206f49607561f9eeeea1545fe3d29e1644acd22b57R1-R27) `lib/features/academic_calendar/model/day_swap_model.dart` [[2]](diffhunk://#diff-46fabfbb0f8945530739dc6b636feb5ef4edd95eb9bb43f1aa71433a25429306R1-R26)
* **Repository Update**: Replaced GraphQL queries with REST API calls in `academicCalendarRepo`, introducing `AcademicCalendarWithSwaps` to combine calendar data and day swaps. (`lib/features/academic_calendar/repository/academic_calendar_repo.dart` [lib/features/academic_calendar/repository/academic_calendar_repo.dartL4-R56](diffhunk://#diff-a50f393a402bb085a0d13345be2291faa7b6c6b6a80a5fb72ff05e353531fa22L4-R56))
* **Removed GraphQL Query**: Deleted the `GetAcademicCalendar` GraphQL query, as it is no longer used. (`lib/features/academic_calendar/repository/getAcademicCalendar.graphql` [lib/features/academic_calendar/repository/getAcademicCalendar.graphqlL1-L13](diffhunk://#diff-61fc153dd7526fc2a92565119e2d4ca7ac80e993d27cb103a0b5a7c5bc9940c2L1-L13))

### Refactoring and Adjustments for New Data Model:
* **Extension Updates**: Updated extensions to work with `AcademicCalendarWithSwaps` instead of the old GraphQL models. (`lib/features/academic_calendar/model/academic_calendar_extensions.dart` [[1]](diffhunk://#diff-9de51e26f447c87597ca9c5a345221725b50876a19aa6974295ec5064087ade5R3-R14) [[2]](diffhunk://#diff-9de51e26f447c87597ca9c5a345221725b50876a19aa6974295ec5064087ade5L39-R45) `lib/features/academic_calendar/utils/counter_digits.dart` [[3]](diffhunk://#diff-76a406ba698b4ea5706a23c5e62a235d502c7daf927a724c7ff50290508a48c1L10-R19) `lib/features/academic_calendar/utils/localize_counter_message.dart` [[4]](diffhunk://#diff-3dfeb5d80a6a07777f94bfe9b89db4f07855960e17c24b91f2461621cdb0e813L7-L16)
* **Widget Adjustments**: Updated widgets like `AcademicCalendarConsumer`, `ExamSessionCountdown`, and `DigitsRow` to use the new `AcademicCalendarWithSwaps` model. (`lib/features/academic_calendar/widgets/academic_calendar_consumer.dart` [[1]](diffhunk://#diff-e1ce6bdd834baffcba0a1398ca5fb644dcc437ddb69023d4e6da5bbcb4cd078dL23-R30) `lib/features/academic_calendar/widgets/countdown_widget/digits_widgets.dart` [[2]](diffhunk://#diff-ca2d5d6ca034a5179665b935691dc5f7084cebd1a104366ba8c91ff777c4518bL10-R10) `lib/features/academic_calendar/widgets/countdown_widget/exam_session_countdown.dart` [[3]](diffhunk://#diff-fa0977bd66502e41be62dbd769f8ff166421d89a151a6c83ce15cc49350902f0L15-R15)

### Dependency Updates:
* **Dependency Version Update**: Updated the `com.android.application` plugin version from `8.3.2` to `8.6.0` in `android/settings.gradle`.
* **Removed Unused Dependency**: Removed `svg_optimizer` from `pubspec.yaml` as it is no longer required. (`pubspec.yaml` [[1]](diffhunk://#diff-8b7e9df87668ffa6a04b32e1769a33434999e54ae081c52e5d943c541d4c0d25L155) [[2]](diffhunk://#diff-8b7e9df87668ffa6a04b32e1769a33434999e54ae081c52e5d943c541d4c0d25L174-L175)

### Minor Enhancements:
* **Greeting Widget Update**: Added a static method to the `Greeting` widget for localized offline error messages and adjusted its `academicCalendar` property to use the new model. (`lib/features/academic_calendar/widgets/home_screen_greeting.dart` [[1]](diffhunk://#diff-9d74e40eadd575292502f5687e3c3fb20052e5882e704c161dd7407942b6c15aR5-R16) [[2]](diffhunk://#diff-9d74e40eadd575292502f5687e3c3fb20052e5882e704c161dd7407942b6c15aL21-R26)

These changes improve the maintainability and scalability of the academic calendar feature by adopting a more robust REST-based architecture and simplifying the data model.
